src/app.rs

```diff
@@ -476,6 +479,144 @@ impl App {
+            None => {
+                panic!("this should not happen?");
+            }
+        }
+
+        // Check tree root
+        if let Err(error) = self.contracts.assert_valid_root(tree.root()).await {
+            error!(
+                computed_root = ?tree.root(),
+                ?error,
+                "Root mismatch between tree and contract."
+            );
+            panic!("Root mismatch between tree and contract.");
+        }
+
+        // Immediately write the tree to storage, before anyone else can
+        // TODO: Store tree in database
+
+        Ok(())
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn spawn_identity_committer(self: Arc<Self>) {
+        let mut listener = self.event_bus.subscribe();
+        tokio::spawn(async move {
+            let mut transaction_in_progress = false;
+            let (tx, mut rx) = tokio::sync::mpsc::channel(1);
+            let mut work_queue: VecDeque<(usize, Hash)> = VecDeque::new();
+            loop {
+                if !transaction_in_progress && !work_queue.is_empty() {
+                    let identity = work_queue
+                        .pop_front()
+                        .expect("Impossible, queue is checked non-empty");
+                    transaction_in_progress = true;
+                    let tx = tx.clone();
+                    let app = self.clone();
+                    tokio::spawn(async move {
+                        info!("Committing identity {:?}", identity);
+                        match app.commit_identity(identity.0, identity.1).await {
+                            Ok(_) => {
+                                info!("Identity committed.");
+                                tx.send(()).await.expect(
+                                    "Failed to send completion signal, identity committer is \
+                                     likely dead, terminating.",
+                                );
+                            }
+                            Err(e) => {
+                                error!(?e, "Failed to commit identity.");
+                                panic!("Failed to commit identity.");
+                            }
+                        }
+                    });
+                }
+
+                select! {
+                    _ = rx.recv() => {
+                        info!("Identity committer job finished.");
+                        transaction_in_progress = false;
+                    },
+                    Ok(msgs) = listener.recv() => {
+                        for msg in msgs {
+                            let Event::PendingIdentityInserted { group_id, commitment } = msg;
+                            info!("Identity committer received event group_id: {}, identity: {}.", group_id, commitment);
+                            work_queue.push_back((group_id, commitment));
+                        }
+                    },
+                    else => {}
+                }
+            }
+        });
```

Comment:
Why not

```
while let Some(event) = listener.recv().await {
 app.commit_identity(...
}
```
?

Reply:
Or even simpler, just use the database as work queue. And poll the database in a loop.

To make it reactive we can have it stop when there is no more work and use an event to start the loop. This can be a simple [Semaphore](https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html) and doesn't have to carry state like the event_bus.